File: 1249.minimum-remove-to-make-valid-parentheses.cpp
/*
 * @lc app=leetcode id=1249 lang=cpp
 *
 * [1249] Minimum Remove to Make Valid Parentheses
 */
#include <algorithm>
#include <stack>
#include <string>

using namespace std;

// @lc code=start
class Solution {
public:
    /*
    // Solution 1
    // T: O(n)
    // S: O(n)
    string minRemoveToMakeValid(string s) {
        const int n = s.size();
        bool keep_flags[n];
        fill_n(keep_flags, n, true);

        stack<int> stk;
        for (int i = 0; i < n; ++i) {
            if (s[i] == '(') {
                // not match a ')' yet
                stk.push(i);
                keep_flags[i] = false;
            } else if (s[i] == ')') {
                if (stk.empty()) { 
                    // cannot match any '('
                    keep_flags[i] = false;
                } else {
                    int pre_idx = stk.top(); stk.pop();
                    keep_flags[pre_idx] = true;
                }
            }
        }

        string res;
        for (int i = 0; i < n; ++i) {
            if (keep_flags[i])  res.push_back(s[i]);
        }
        return res;
    }
    */

    // Solution 2: in-place modification
    // T: O(n)
    // S: O(n)
#define REMOVE_SIGN '*'
    string minRemoveToMakeValid(string s) {
        stack<int> stk;
        for (int i = 0; i < s.size(); ++i) {
            if (s[i] == '(') {
                stk.push(i);
            } else if (s[i] == ')') {
                if (stk.empty()) {
                    // Could not match any '('
                    s[i] = REMOVE_SIGN;
                } else {
                    stk.pop();
                }
            }
        }

        while (stk.size()) {
            // Find any '(' without matching ')'
            s[stk.top()] = REMOVE_SIGN; stk.pop();
        }

        s.erase(remove(s.begin(), s.end(), REMOVE_SIGN), s.end());
        return s;
    }
};
// @lc code=end

File: 1249.minimum-remove-to-make-valid-parentheses.py
#
# @lc app=leetcode id=1249 lang=python3
#
# [1249] Minimum Remove to Make Valid Parentheses
#

# @lc code=start
class Solution:
    # T: O(n)
    # S: O(n)
    def minRemoveToMakeValid(self, s: str) -> str:
        N = len(s)
        keep_mask = [True] * N

        stack = []
        for i, c in enumerate(s):
            if c == '(':
                # not yet matched to a ')'
                stack.append(i)
                keep_mask[i] = False
            elif c == ')':
                if len(stack) == 0:
                    # can't be matched to a previous '('
                    keep_mask[i] = False
                else:
                    last_left_idx = stack.pop()
                    keep_mask[last_left_idx] = True

        return ''.join(c for i, c in enumerate(s) if keep_mask[i])

        
# @lc code=end

